modules/natgw/main.tf

```
@@ -0,0 +1,96 @@
+resource "azurerm_public_ip" "this" {
+  count = (var.create_natgw && var.create_pip) ? 1 : 0
+
+  name                = "${var.name}-pip"
+  resource_group_name = var.resource_group_name
+  location            = var.location
+  allocation_method   = "Static"
+  sku                 = "Standard"
+  zones               = var.zone != null ? [var.zone] : null
+
+  tags = var.tags
+}
+
+data "azurerm_public_ip" "this" {
+  count = (var.create_natgw && !var.create_pip && var.existing_pip_name != null) ? 1 : 0
+
+  name                = var.existing_pip_name
+  resource_group_name = var.existing_pip_resource_group_name == null ? var.resource_group_name : var.existing_pip_resource_group_name
+}
+
+resource "azurerm_public_ip_prefix" "this" {
+  count = (var.create_natgw && var.create_pip_prefix) ? 1 : 0
+
+  name                = "${var.name}-pip-prefix"
+  resource_group_name = var.resource_group_name
+  location            = var.location
+  ip_version          = "IPv4"
+  prefix_length       = var.pip_prefix_length
+  sku                 = "Standard"
+  zones               = var.zone != null ? [var.zone] : null
+
+  tags = var.tags
+}
+
+data "azurerm_public_ip_prefix" "this" {
+  count = (var.create_natgw && !var.create_pip_prefix && var.existing_pip_prefix_name != null) ? 1 : 0
```

In case, when ``create_pip_prefix`` is false and ``existing_pip_prefix_name`` is null, we just use ``try(data.azurerm_public_ip_prefix.this[0], null)`` as in line 73 ? Maybe we could simplify it by making ``create_pip_prefix`` local which is true only if variable ``existing_pip_prefix_name`` is empty ?

---

1. I do not follow :) you would have to drop a code suggestion
2. having this logic here gives two benefits:
  * you call this data source only when you need to
  * this condition documents itself (although it is hard to read ;))

If I understood correctly you want to simplify the condition and move part of the logic to a `try` statement further in the code. But this would brake both bullet points.

---

I'm sorry for unclear description. I was mainly thinking about simplifying and making it easier to understand that condition by use local - then it's ready to catch that we are using data source if we are creating NAT GW and we want to use public IP prefix e.g.

```suggestion
  count = (var.create_natgw && local.use_pip_prefix) ? 1 : 0
```

where local is true on following condition:

```
locals {
  use_pip_prefix = !var.create_pip_prefix && var.existing_pip_prefix_name != null
}
```

---

I would stick to the original code. We only split the logic into to places. If someone really wants to understand what is happening, one has to go to `locals` to figure it out.